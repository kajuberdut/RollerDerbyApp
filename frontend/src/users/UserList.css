.UserList {
    background-color: #d1d2d4;
    margin-top: 150px;
    width: 43%;
    margin-left: 12%;
}