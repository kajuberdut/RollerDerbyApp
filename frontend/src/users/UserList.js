import React, { useState, useEffect } from "react";
import FastApi from "../Api";
import UserComponent from "./UserComponent";
import "./UserList.css"
import Loading from "../multiUse/loading/Loading"
import SearchComponentUsers from "../multiUse/searchComponent/SearchComponentUsers";
import { Card, CardBody, CardTitle } from "reactstrap";

/**
 * Display users page
 */

function UserList() {

    /** Set Users and is loading in state*/

    const [isLoading, setIsLoading] = useState(true);
    const [users, setUsers] = useState([]);
    const user = JSON.parse(localStorage.getItem('user'));

  /** API get request for users */

  async function getUsers() {

    try {
      let users = await FastApi.getUsers();
      console.log("users in userlist !!!!!!!!!!!!!!!  ", users)
      setUsers(users)
      setIsLoading(false)
    } catch (errors) {
      return { success: false, errors };
    }
  }

   /** Reloads users when changes request for users */

    useEffect(() => {
        getUsers();
        console.log("users!!!!!!!", users)
    }, []);


  /** Display loading if API call is has not returned */

    if (isLoading) {
      return (
          <Loading />
      )
    }

  /** Render the cards for users*/

    const renderCards = () => {
      return (
        <div className="UserList-RenderCards">
                {users.map(indUser => (
                  indUser.username !== user.username ? (
                  <UserComponent indUser={indUser} key={"User-" + indUser.userId} />
                  ) : null 
                ))}
          </div>
        );
    }
  
  /** Render search bar and cards */

    return (
      <>
      <div style={{marginTop: '80px', minWidth: '400px'}}>
          <SearchComponentUsers getUsers={getUsers} setUsers={setUsers}/>
      </div>
      <Card className="UserList" style={{minWidth: '400px'}}>
        <CardBody>
        <CardTitle><h1>Users</h1></CardTitle>
        {users && <div> {renderCards()} </div> }
        {!users && <div>Sorry, no users were located.</div>}
        </CardBody>
      </Card>
      </>
    );

}

export default UserList;