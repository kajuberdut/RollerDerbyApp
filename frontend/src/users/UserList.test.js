import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import UserList from './UserList'; 
import FastApi from '../Api'; 
import { BrowserRouter } from "react-router-dom";

// * Passing as of 2/15/24
// npm test UserList.test.js
// must be in frontend directory

// Mock FastApi methods
jest.mock('../Api', () => ({
  getUsers: jest.fn(),
}));


describe('UserList Component', () => {
    const mockUsers = [
      { 
        userId: 1,
        username: "testUser",
        image: null
      },
      { 
          userId: 2,
          username: "User2",
          image: null
      },
      { 
          userId: 3,
          username: "User3",
          image: null
      }
    ]

  beforeEach(() => {
    FastApi.getUsers.mockResolvedValue(mockUsers);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('fetches users and displays page details', async () => {
    render(
    <BrowserRouter>
        <UserList />
    </BrowserRouter>
    );

    expect(screen.getByTestId('loading-indicator')).toBeInTheDocument();

    // Wait for the API call to resolve and check if the details are displayed
    await waitFor(() => {
        expect(screen.getByText('Search')).toBeInTheDocument();
        expect(screen.getByText('Users')).toBeInTheDocument();
    });

  });

  it('fetches users and users names', async () => {
    render(
    <BrowserRouter>
        <UserList />
    </BrowserRouter>
    );

    expect(screen.getByTestId('loading-indicator')).toBeInTheDocument();

    // Wait for the API call to resolve and check if the details are displayed
    await waitFor(() => {
        expect(screen.getByText('User2')).toBeInTheDocument();
        expect(screen.getByText('User3')).toBeInTheDocument();
    });

  });

  it('fetches users and does not show current users name', async () => {
    render(
    <BrowserRouter>
        <UserList />
    </BrowserRouter>
    );

    expect(screen.getByTestId('loading-indicator')).toBeInTheDocument();

    // Wait for the API call to resolve and check if the details are displayed
    await waitFor(() => {
        expect(screen.queryByText('testUser')).not.toBeInTheDocument();
    });

  });


  test('matches snapshot', function() {
    const { asFragment } = render(<UserList />);
    expect(asFragment()).toMatchSnapshot(); 
  })

});

