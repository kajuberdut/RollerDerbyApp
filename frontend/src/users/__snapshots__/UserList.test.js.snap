// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Loading"
    data-testid="loading-indicator"
  />
</DocumentFragment>
`;
